crate/server/src/core/operations/mac.rs

```diff
@@ -0,0 +1,220 @@
+use std::sync::Arc;
+
+use cosmian_kmip::kmip_2_1::{
+    kmip_operations::{Mac, MacResponse},
+    kmip_types::{HashingAlgorithm, UniqueIdentifier},
+    KmipOperation,
+};
+use cosmian_kms_interfaces::SessionParams;
+use openssl::{md::Md, md_ctx::MdCtx, pkey::PKey};
+use tracing::trace;
+
+use crate::{
+    core::{retrieve_object_utils::retrieve_object_for_operation, KMS},
+    error::KmsError,
+    kms_bail,
+    result::{KResult, KResultHelper},
+};
+
+fn compute_hmac(key: &[u8], data: &[u8], algorithm: HashingAlgorithm) -> KResult<Vec<u8>> {
+    let message_digest = match algorithm {
+        HashingAlgorithm::SHA1 => Md::sha1(),
+        HashingAlgorithm::SHA224 => Md::sha224(),
+        HashingAlgorithm::SHA256 => Md::sha256(),
+        HashingAlgorithm::SHA384 => Md::sha384(),
+        HashingAlgorithm::SHA512 => Md::sha512(),
+        HashingAlgorithm::SHA3224 => Md::sha3_224(),
+        HashingAlgorithm::SHA3256 => Md::sha3_256(),
+        HashingAlgorithm::SHA3384 => Md::sha3_384(),
+        HashingAlgorithm::SHA3512 => Md::sha3_512(),
+        algorithm => kms_bail!("Unsupported hashing algorithm: {:?}", algorithm),
+    };
+
+    let key = PKey::hmac(key)?;
+    let mut ctx = MdCtx::new()?;
+    ctx.digest_sign_init(Some(message_digest), &key)?;
+    ctx.digest_sign_update(data)?;
+    let mut hmac = Vec::with_capacity(64); // 512 bits being the maximum size of supported hash functions
+    ctx.digest_sign_final_to_vec(&mut hmac)?;
+    Ok(hmac)
+}
+
+pub(crate) async fn mac(
+    kms: &KMS,
+    request: Mac,
+    user: &str,
+    params: Option<Arc<dyn SessionParams>>,
+) -> KResult<MacResponse> {
+    trace!("Mac: {}", serde_json::to_string(&request)?);
+
+    let unique_identifier = request
+        .unique_identifier
+        .as_ref()
+        .ok_or(KmsError::UnsupportedPlaceholder)?
+        .as_str()
+        .context("Mac: unique_identifier must be a string")?;
+
+    let algorithm = request
+        .cryptographic_parameters
+        .hashing_algorithm
+        .ok_or_else(|| KmsError::InvalidRequest("Hashing algorithm is required".to_owned()))?;
+
+    let data = request
+        .data
+        .as_ref()
+        .ok_or_else(|| KmsError::InvalidRequest("Data is required".to_owned()))?;
+
+    let owm =
+        retrieve_object_for_operation(unique_identifier, KmipOperation::Get, kms, user, params)
+            .await?;
+
+    let unique_identifier = UniqueIdentifier::TextString(unique_identifier.to_owned());
+
+    let hmac = if let Some(correlation_value) = request.correlation_value {
```

Comment:
HMAC is an algorithm, I prefer the term digest, or maybe mac.
```suggestion
    let digest = if let Some(correlation_value) = request.correlation_value {
```